x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/repository/CcrRepository.java

```diff
@@ -237,25 +253,21 @@ public void restoreShard(IndexShard indexShard, SnapshotId snapshotId, Version v
             store.decRef();
         }
 
-        Store.MetadataSnapshot recoveryMetadata;
-        try {
-            recoveryMetadata = indexShard.snapshotStoreMetadata();
-        } catch (IOException e) {
-            throw new IndexShardRecoveryException(shardId, "failed access store metadata", e);
-        }
-
         Map<String, String> ccrMetaData = indexShard.indexSettings().getIndexMetaData().getCustomData(Ccr.CCR_CUSTOM_METADATA_KEY);
         String leaderUUID = ccrMetaData.get(Ccr.CCR_CUSTOM_METADATA_LEADER_INDEX_UUID_KEY);
         Index leaderIndex = new Index(shardId.getIndexName(), leaderUUID);
         ShardId leaderShardId = new ShardId(leaderIndex, shardId.getId());
 
         Client remoteClient = client.getRemoteClusterClient(remoteClusterAlias);
-        String sessionUUID = UUIDs.randomBase64UUID();
-        PutCcrRestoreSessionAction.PutCcrRestoreSessionResponse response = remoteClient.execute(PutCcrRestoreSessionAction.INSTANCE,
-            new PutCcrRestoreSessionRequest(sessionUUID, leaderShardId, recoveryMetadata)).actionGet();
-        DiscoveryNode node = response.getNode();
-        // TODO: Implement file restore
-        closeSession(remoteClient, node, sessionUUID);
+        // TODO: There should be some local timeout. And if the remote cluster returns an unknown session
+        //  response, we should be able to retry by creating a new session.
+        String name = metadata.name();
+        try (RestoreSession restoreSession = RestoreSession.openSession(name, remoteClient, leaderShardId, indexShard, recoveryState)) {
+            restoreSession.restoreFiles();
```

Comment:
out of curiosity, do we plan do to this async at some point? given that this is a new development what is the rational to do it sync? It's because snapshot is sync?

Reply:
Yes snapshot is sync, which is why this is synchronous. Theoretically it could be async, but I imagine that would be a larger change and would have required us to construct a lot of the snapshot/restore stuff that already exists.

Reply:
++